frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+			call.clone().dispatch(origin).map_err(|e| e.error)?;
+
+			let _ = Self::do_buy_ticket(&caller, &call);
+		}
+
+		#[weight = T::WeightInfo::set_calls(calls.len() as u32)]
+		fn set_calls(origin, calls: Vec<<T as Config>::Call>) {
+			T::ManagerOrigin::ensure_origin(origin)?;
+			ensure!(calls.len() <= T::MaxCalls::get(), Error::<T>::TooManyCalls);
+			if calls.is_empty() {
+				CallIndices::kill();
+			} else {
+				let indices = Self::calls_to_indices(&calls)?;
+				CallIndices::put(indices);
+			}
+			Self::deposit_event(RawEvent::CallsUpdated);
+		}
+
+		/// Start a lottery using the provided configuration.
+		///
+		/// This extrinsic must be called by the `ManagerOrigin`.
+		///
+		/// Parameters:
+		///
+		/// * `price`: The cost of a single ticket.
+		/// * `length`: How long the lottery should run for starting at the current block.
+		/// * `delay`: How long after the lottery end we should wait before picking a winner.
+		/// * `repeat`: If the lottery should repeat when completed.
+		#[weight = T::WeightInfo::start_lottery()]
+		fn start_lottery(origin,
+			price: BalanceOf<T>,
+			length: T::BlockNumber,
+			delay: T::BlockNumber,
+			repeat: bool,
+		) {
+			T::ManagerOrigin::ensure_origin(origin)?;
+			Lottery::<T>::try_mutate(|lottery| -> DispatchResult {
+				ensure!(lottery.is_none(), Error::<T>::InProgress);
+				let index = LotteryIndex::get();
+				let new_index = index.checked_add(1).ok_or(Error::<T>::Overflow)?;
+				let start = frame_system::Module::<T>::block_number();
+				// Use new_index to more easily track everything with the current state.
+				*lottery = Some(LotteryConfig {
+					price,
+					start,
+					length,
+					delay,
+					repeat,
+				});
+				LotteryIndex::put(new_index);
+				Ok(())
+			})?;
+			Self::deposit_event(RawEvent::LotteryStarted);
+		}
+
+		/// If a lottery is repeating, you can use this to stop the repeat.
+		/// The lottery will continue to run to completion.
+		///
+		/// This extrinsic must be called by the `ManagerOrigin`.
+		#[weight = T::WeightInfo::stop_repeat()]
+		fn stop_repeat(origin) {
+			T::ManagerOrigin::ensure_origin(origin)?;
+			Lottery::<T>::mutate(|mut lottery| {
+				if let Some(config) = &mut lottery {
+					config.repeat = false
+				}
+			});
+		}
+
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			Lottery::<T>::mutate(|mut lottery| -> Weight {
+				if let Some(config) = &mut lottery {
+					let payout_block = config.start
+						.saturating_add(config.length)
+						.saturating_add(config.delay);
+					if payout_block <= n {
+						let (lottery_account, lottery_balance) = Self::pot();
+						let ticket_count = TicketsCount::get();
+
+						let winning_number = Self::choose_winner(ticket_count);
+						let winner = Tickets::<T>::get(winning_number).unwrap_or(lottery_account);
+						// Not much we can do if this fails...
+						let _ = T::Currency::transfer(&Self::account_id(), &winner, lottery_balance, KeepAlive);
```

Comment:
I don't see where account_id is initialized, maybe we should add a genesis_config ? or just do it on start_lottery ?

Otherwise this fails when price is less than existential deposit.

Also this could be done by telling the account has one provider once https://github.com/paritytech/substrate/pull/7363 is merged